### PR TITLE
Revert "Synchronize stream in `DeviceBuffer` only on copy"

### PR DESCRIPTION
**Pull request by jakirkham**
Appears PR ( https://github.com/rapidsai/rmm/pull/326 ) resulted in segfaults when using cuDF with UCX that appears to be caused by a race condition. Given this, go ahead and revert that change for now to unblock users until we can come up with something better.

cc @kkraus14 @jrhemstad @pentschev @beckernick

### CODE REVIEW

**Comment by jakirkham** [comment]
cc @quasiben

**Comment by quasiben** [comment]
As an experiment should we add `numba.cuda. synchronize()` calls in the send in dask ?

**Comment by pentschev** [comment]
> As an experiment should we add numba.cuda. synchronize() calls in the send in dask ?

Unfortunately, that didn't work either. The only thing that did work so far was to synchronize the entire context just before this line: https://github.com/rapidsai/ucx-py/blob/branch-0.13/ucp/_libs/core.pyx#L728 .

I tried `numba.cuda.synchronize()` at the same spot, as well as after that and after reading, and I still get segfaults. I'm not sure if that synchronizes the default stream, and given `numba.cuda.default_stream()` isn't available in Numba 0.48, I also tried ensuring I was indeed synchronizing the default stream as follows:

```python
Stream = numba.cuda.driver.Stream
ctx = numba.cuda.current_context()
stream = Stream(ctx, numba.cuda.driver.drvapi.cu_stream(0), None)
stream.synchronize()
```

Adding context or stream synchronizations to read/write functions in distributed/comm/ucx.py also doesn't get rid of the segfault.

To summarize, the working solutions for https://github.com/rapidsai/tpcx-bb/issues/378 so far are:

1. This PR;
2. Synchronizing the context in `ucp._libs.core._Endpoint.send`.

It's important to note that both solutions above fix https://github.com/rapidsai/tpcx-bb/issues/378 but _do not fix InfiniBand_ issues, we still get hangs/crashes even having both at the same time plus `CUDA_LAUNCH_BLOCKING=1`. Therefore, I'm thinking now that maybe IB issues are not due to synchronization after all.

**Comment by pentschev** [comment]
> As an experiment should we add numba.cuda. synchronize() calls in the send in dask ?

Unfortunately, that didn't work either. The only thing that did work so far was to synchronize the entire context just before this line: https://github.com/rapidsai/ucx-py/blob/branch-0.13/ucp/_libs/core.pyx#L728 .

I tried `numba.cuda.synchronize()` at the same spot, as well as after that and after reading, and I still get segfaults. I'm not sure if that synchronizes the default stream, and given `numba.cuda.default_stream()` isn't available in Numba 0.48, I also tried ensuring I was indeed synchronizing the default stream as follows:

```python
Stream = numba.cuda.driver.Stream
ctx = numba.cuda.current_context()
stream = Stream(ctx, numba.cuda.driver.drvapi.cu_stream(0), None)
stream.synchronize()
Adding context or stream synchronizations to read/write functions in distributed/comm/ucx.py also doesn't get rid of the segfault.
```

To summarize, the working solutions so far are:

1. This PR;
1. Synchronizing the context in ucp._libs.core._Endpoint.send.

It's important to note that both solutions above fix the issue but do not fix InfiniBand issues, we still get hangs/crashes even having both at the same time plus CUDA_LAUNCH_BLOCKING=1. Therefore, I'm thinking now that maybe IB issues are not due to synchronization after all.

**Comment by jakirkham** [comment]
Addressed by synchronizing in Dask before sending and after allocating memory before receiving ( https://github.com/dask/distributed/pull/3598 ). Closing this out.
